File: packages/widget-bridge/android/src/main/java/com/todoapp/widgetbridge/WidgetBridgeModule.java
package com.todoapp.widgetbridge;

import android.appwidget.AppWidgetManager;
import android.content.ComponentName;
import android.content.Context;
import android.content.Intent;
import android.content.SharedPreferences;

import com.facebook.react.bridge.Promise;
import com.facebook.react.bridge.ReactApplicationContext;
import com.facebook.react.bridge.ReactContextBaseJavaModule;
import com.facebook.react.bridge.ReactMethod;
import com.facebook.react.bridge.ReadableMap;
import com.facebook.react.bridge.ReadableArray;
import com.facebook.react.module.annotations.ReactModule;

import org.json.JSONArray;
import org.json.JSONObject;

@ReactModule(name = "WidgetBridgeModule")
public class WidgetBridgeModule extends ReactContextBaseJavaModule {
    private static final String MODULE_NAME = "WidgetBridgeModule";
    private static final String PREFS_NAME = "TodoWidgetPrefs";
    private static final String TASKS_KEY = "tasks_data";

    public WidgetBridgeModule(ReactApplicationContext reactContext) {
        super(reactContext);
    }

    @Override
    public String getName() {
        return MODULE_NAME;
    }

    @ReactMethod
    public void updateWidget(ReadableMap data, Promise promise) {
        try {
            Context context = getReactApplicationContext();
            SharedPreferences prefs = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE);
            SharedPreferences.Editor editor = prefs.edit();

            // Convert ReadableMap to JSON
            JSONObject jsonData = new JSONObject();
            
            if (data.hasKey("tasks")) {
                ReadableArray tasksArray = data.getArray("tasks");
                JSONArray jsonTasks = new JSONArray();
                
                for (int i = 0; i < tasksArray.size(); i++) {
                    ReadableMap task = tasksArray.getMap(i);
                    JSONObject jsonTask = new JSONObject();
                    
                    jsonTask.put("id", task.getString("id"));
                    jsonTask.put("title", task.getString("title"));
                    jsonTask.put("completed", task.getBoolean("completed"));
                    
                    if (task.hasKey("dueDate")) {
                        jsonTask.put("dueDate", task.getString("dueDate"));
                    }
                    
                    if (task.hasKey("priority")) {
                        jsonTask.put("priority", task.getString("priority"));
                    }
                    
                    jsonTasks.put(jsonTask);
                }
                
                jsonData.put("tasks", jsonTasks);
            }
            
            if (data.hasKey("lastUpdated")) {
                jsonData.put("lastUpdated", data.getString("lastUpdated"));
            }

            // Save data to SharedPreferences
            editor.putString(TASKS_KEY, jsonData.toString());
            editor.apply();

            // Trigger widget update
            updateAllWidgets(context);
            
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("UPDATE_WIDGET_ERROR", "Failed to update widget", e);
        }
    }

    @ReactMethod
    public void refreshWidget(Promise promise) {
        try {
            Context context = getReactApplicationContext();
            updateAllWidgets(context);
            promise.resolve(null);
        } catch (Exception e) {
            promise.reject("REFRESH_WIDGET_ERROR", "Failed to refresh widget", e);
        }
    }

    @ReactMethod
    public void isWidgetSupported(Promise promise) {
        try {
            Context context = getReactApplicationContext();
            AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
            
            // Check if any widgets are installed
            ComponentName provider = new ComponentName(context.getPackageName(), 
                context.getPackageName() + ".TodoWidgetProvider");
            int[] widgetIds = appWidgetManager.getAppWidgetIds(provider);
            
            promise.resolve(widgetIds.length > 0);
        } catch (Exception e) {
            promise.resolve(false);
        }
    }

    private void updateAllWidgets(Context context) {
        Intent intent = new Intent(context.getPackageName() + ".UPDATE_WIDGET");
        intent.setPackage(context.getPackageName());
        context.sendBroadcast(intent);
        
        // Also use AppWidgetManager to force update
        AppWidgetManager appWidgetManager = AppWidgetManager.getInstance(context);
        ComponentName provider = new ComponentName(context.getPackageName(), 
            context.getPackageName() + ".TodoWidgetProvider");
        int[] widgetIds = appWidgetManager.getAppWidgetIds(provider);
        
        if (widgetIds.length > 0) {
            Intent updateIntent = new Intent(AppWidgetManager.ACTION_APPWIDGET_UPDATE);
            updateIntent.putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, widgetIds);
            updateIntent.setComponent(provider);
            context.sendBroadcast(updateIntent);
        }
    }
}

File: packages/widget-bridge/android/src/main/kotlin/com/productivity/widgetbridge/WidgetBridgeModule.kt
package com.productivity.widgetbridge

import com.facebook.react.bridge.*
import com.facebook.react.modules.core.DeviceEventManagerModule
import android.content.Intent
import android.content.Context
import android.net.Uri
import android.content.SharedPreferences
import android.content.ComponentName
import android.appwidget.AppWidgetManager
import org.json.JSONObject

class WidgetBridgeModule(reactContext: ReactApplicationContext) : 
    ReactContextBaseJavaModule(reactContext) {

    companion object {
        const val MODULE_NAME = "WidgetBridge"
        const val PREFS_NAME = "ProductivityWidgetPrefs"
        const val PREF_WIDGET_DATA = "widget_data_"
    }

    override fun getName() = MODULE_NAME

    private fun getSharedPreferences(): SharedPreferences {
        return reactApplicationContext.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
    }

    @ReactMethod
    fun updateWidgetData(widgetId: String, dataJson: String, promise: Promise) {
        try {
            val prefs = getSharedPreferences()
            prefs.edit().putString(PREF_WIDGET_DATA + widgetId, dataJson).apply()
            
            // Trigger widget update
            triggerWidgetUpdate()
            
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("UPDATE_FAILED", e.message)
        }
    }

    @ReactMethod
    fun getWidgetData(widgetId: String, promise: Promise) {
        try {
            val prefs = getSharedPreferences()
            val data = prefs.getString(PREF_WIDGET_DATA + widgetId, null)
            promise.resolve(data)
        } catch (e: Exception) {
            promise.reject("GET_FAILED", e.message)
        }
    }

    @ReactMethod
    fun scheduleUpdate(widgetId: String, delayMs: Double, promise: Promise) {
        try {
            // Use the widget update service
            val intent = Intent("com.productivity.widget.UPDATE_WIDGETS").apply {
                putExtra("widgetId", widgetId)
                putExtra("delayMs", delayMs.toLong())
            }
            reactApplicationContext.sendBroadcast(intent)
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("SCHEDULE_FAILED", e.message)
        }
    }

    @ReactMethod
    fun forceUpdateAll(promise: Promise) {
        try {
            triggerWidgetUpdate()
            promise.resolve(null)
        } catch (e: Exception) {
            promise.reject("UPDATE_FAILED", e.message)
        }
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun supportsInteraction(): Boolean {
        // Android widgets support limited interaction
        return true
    }

    @ReactMethod(isBlockingSynchronousMethod = true)
    fun canCompleteTask(): Boolean {
        // Android widgets can complete tasks via checkbox
        return true
    }

    @ReactMethod
    fun openDeepLink(url: String, params: String?, promise: Promise) {
        try {
            val intent = Intent(Intent.ACTION_VIEW, Uri.parse(url)).apply {
                flags = Intent.FLAG_ACTIVITY_NEW_TASK
                if (params != null) {
                    putExtra("params", params)
                }
            }
            
            if (intent.resolveActivity(reactApplicationContext.packageManager) != null) {
                reactApplicationContext.startActivity(intent)
                promise.resolve(null)
            } else {
                promise.reject("NO_ACTIVITY", "No activity found to handle the deep link")
            }
        } catch (e: Exception) {
            promise.reject("DEEPLINK_FAILED", e.message)
        }
    }

    private fun triggerWidgetUpdate() {
        val intent = Intent(reactApplicationContext, getWidgetProviderClass()).apply {
            action = AppWidgetManager.ACTION_APPWIDGET_UPDATE
            val appWidgetManager = AppWidgetManager.getInstance(reactApplicationContext)
            val widgetComponent = ComponentName(
                reactApplicationContext.packageName,
                "com.productivity.widget.TodoWidgetProvider"
            )
            val ids = appWidgetManager.getAppWidgetIds(widgetComponent)
            putExtra(AppWidgetManager.EXTRA_APPWIDGET_IDS, ids)
        }
        reactApplicationContext.sendBroadcast(intent)
    }

    private fun getWidgetProviderClass(): Class<*> {
        return Class.forName("com.productivity.widget.TodoWidgetProvider")
    }

    // Send events to JavaScript
    private fun sendEvent(eventName: String, params: WritableMap) {
        reactApplicationContext
            .getJSModule(DeviceEventManagerModule.RCTDeviceEventEmitter::class.java)
            .emit(eventName, params)
    }

    // Listen for widget events
    fun onWidgetTaskCompleted(taskId: String, completed: Boolean) {
        val params = Arguments.createMap().apply {
            putString("type", "TASK_COMPLETED")
            putString("taskId", taskId)
            putBoolean("completed", completed)
        }
        sendEvent("WidgetEvent", params)
    }
}